test/OpenTelemetry.Tests/Metrics/MetricAPITest.cs

```diff
@@ -235,20 +237,81 @@ void ProcessExport(IEnumerable<Metric> batch)
+            var metricItems = new List<Metric>();
+            var metricExporter = new TestMetricExporter(ProcessExport);
+
+            void ProcessExport(IEnumerable<Metric> batch)
             {
-                foreach (var metricPoint in metric.GetMetricPoints())
+                foreach (var metricItem in batch)
                 {
-                    sumReceived += metricPoint.LongValue;
+                    metricItems.Add(metricItem);
                 }
             }
 
-            var expectedSum = deltaValueUpdatedByEachCall * numberOfMetricUpdateByEachThread * numberOfThreads;
-            Assert.Equal(expectedSum, sumReceived);
+            var metricReader = new BaseExportingMetricReader(metricExporter);
+
+            var meter = new Meter("TestMeter");
+            var counterDouble = meter.CreateCounter<double>("mycounter");
+            var meterProvider = Sdk.CreateMeterProviderBuilder()
+                .AddSource("TestMeter")
+                .AddMetricReader(metricReader)
+                .Build();
+
+            // setup args to threads.
+            var mreToBlockUpdateThreads = new ManualResetEvent(false);
+            var mreToEnsureAllThreadsStarted = new ManualResetEvent(false);
+
+            var argToThread = new UpdateThreadArguments<double>();
+            argToThread.DeltaValueUpdatedByEachCall = deltaDoubleValueUpdatedByEachCall;
+            argToThread.Counter = counterDouble;
+            argToThread.ThreadsStartedCount = 0;
+            argToThread.MreToBlockUpdateThread = mreToBlockUpdateThreads;
+            argToThread.MreToEnsureAllThreadsStart = mreToEnsureAllThreadsStarted;
+
+            Thread[] t = new Thread[numberOfThreads];
+            for (int i = 0; i < numberOfThreads; i++)
+            {
+                t[i] = new Thread(CounterUpdateThread<double>);
+                t[i].Start(argToThread);
+            }
+
+            // Block until all threads started.
+            mreToEnsureAllThreadsStarted.WaitOne();
+
+            Stopwatch sw = new Stopwatch();
+            sw.Start();
+
+            // unblock all the threads.
+            // (i.e let them start counter.Add)
+            mreToBlockUpdateThreads.Set();
+
+            for (int i = 0; i < numberOfThreads; i++)
+            {
+                // wait for all threads to complete
+                t[i].Join();
+            }
+
+            var timeTakenInMilliseconds = sw.ElapsedMilliseconds;
+            this.output.WriteLine($"Took {timeTakenInMilliseconds} msecs. Total threads: {numberOfThreads}, each thread doing {numberOfMetricUpdateByEachThread} recordings.");
+
+            meterProvider.Dispose();
+            metricReader.Collect();
+
+            var sumReceived = GetDoubleSum(metricItems);
+            var expectedSum = deltaDoubleValueUpdatedByEachCall * numberOfMetricUpdateByEachThread * numberOfThreads;
+            var difference = Math.Abs(sumReceived - expectedSum);
+            Assert.True(difference <= 0.0001);
```

Comment:
If there is a better way to do this, please let me know! :)

Reply:
Assert.Equal(expectedSum, sumReceived, 2);

Reply:
you can do this as follow up PRs. You can consider refactoring to extract common code to helper functions as well.